File: ConstructBinaryTreeFromInorderAndPostorderTraversal/ConstructBinaryTreeFromInorderAndPostorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int, int> inorder_map;
        for(int i = 0; i < inorder.size(); i++) {
        	inorder_map[inorder[i]] = i;
        }
        TreeNode* root = BFS(0, inorder.size() - 1, postorder, 0, postorder.size() - 1, inorder_map);
        return root;
    }
private:
	TreeNode* BFS(int i_begin, int i_end, vector<int>& postorder, int p_begin, int p_end, unordered_map<int, int>& inorder_map) {
		if(i_begin > i_end || p_begin > p_end)
			return nullptr;
		TreeNode* node = new TreeNode(postorder[p_end]);
		int index = inorder_map.find(postorder[p_end])->second;
		int l = index - i_begin;
		node->left = BFS(i_begin, i_begin + l - 1, postorder, p_begin, p_begin + l - 1, inorder_map);
		node->right = BFS(i_begin + l + 1, i_end, postorder, p_begin + l, p_end - 1, inorder_map);
		return node;
	}
};

File: ConstructBinaryTreeFromInorderAndPostorderTraversal/ConstructBinaryTreeFromInorderAndPostorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        # if len(inorder) == 0:
        #     return None
        # root_val = postorder.pop()
        # mid = 0
        # for i in range(len(inorder)):
        #     if inorder[i] == root_val:
        #         mid = i
        #         break
        # root = TreeNode(root_val)
        # root.right = self.buildTree(inorder[mid + 1:], postorder)
        # root.left = self.buildTree(inorder[: mid], postorder)
        # return root
        if len(inorder) == 0:
            return None
        ileft = 0
        iright = len(inorder) - 1
        pleft = 0
        pright = len(postorder) - 1
        return build(inorder, postorder, ileft, iright, pleft, pright)

def build(inorder, postorder, ileft, iright, pleft, pright):
    if ileft == iright:
        return TreeNode(inorder[ileft])
    if ileft > iright or pleft > pright:
        return
    root_val = postorder[pright]
    root = TreeNode(root_val)
    index = inorder.index(root_val)
    index = index - ileft
    root_left = build(inorder, postorder, ileft, ileft + index - 1, pleft, pleft + index - 1)
    root_right = build(inorder, postorder, ileft + index + 1, iright, pleft + index, pright - 1)
    root.left = root_left
    root.right = root_right
    return root